ee/fe/modules/500-basic-auth/hooks/generate_passwords.go

```diff
@@ -6,40 +6,146 @@ Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https
+	}
+
+	return secret.Data, nil
+}
+
+const defaultUserName = `admin`
+
+func defaultLocationValues(passwd string) []map[string]interface{} {
+	return []map[string]interface{}{
+		{
+			"users": map[string]interface{}{
+				defaultUserName: passwd,
+			},
+			"location": "/",
+		},
+	}
+}
 
 func generatePassword(input *go_hook.HookInput) error {
-	_, ok := input.Values.GetOk("basicAuth.locations")
+	// Set values from user controlled configuration.
+	userLocations, ok := input.ConfigValues.GetOk(locationsKey)
 	if ok {
+		input.Values.Set(locationsInternalKey, userLocations.Value())
 		return nil
 	}
 
-	rawLocations := make([]map[string]interface{}, 0)
+	// No config values. Try to restore generated password from the Secret.
+	// Generate default location if no valid generated password available.
 
-	locations := fmt.Sprintf(defaultLocationTemplate, pwgen.AlphaNum(20))
-	err := json.Unmarshal([]byte(locations), &rawLocations)
+	pass, err := restorePasswordFromSnapshot(input.Snapshots[secretBinding])
 	if err != nil {
-		return err
+		input.LogEntry.Infof("Generate default location for basic auth: %s", err)
+		pass = pwgen.AlphaNum(generatedPasswdLength)
 	}
 
-	input.ConfigValues.Set("basicAuth.locations", rawLocations)
+	locations := defaultLocationValues(pass)
+	input.Values.Set(locationsInternalKey, locations)
 	return nil
 }
+
+// restorePasswordFromSnapshot returns generate password from Secret.
+// password is considered generated if:
+// - there is only 1 snapshot
+// - there is only htpasswd field in Secret
+// - there is only one line contains "admin:{PLAIN}" in htpasswd
+// Hook should generate new default location if Secret
+// contains more fields or passwords.
+//
+// NOTE: This algorithm is coupled with the field name in secret.yaml and "users" template in _helpers.tpl.
+func restorePasswordFromSnapshot(snapshot []go_hook.FilterResult) (string, error) {
+	// Only one Secret is expected.
+	if len(snapshot) != 1 {
+		return "", fmt.Errorf("no Secret")
+	}
+
+	secretData, ok := snapshot[0].(map[string][]byte)
+	if !ok {
+		return "", fmt.Errorf("secret has empty data")
+	}
+	// Only one field is expected.
+	if len(secretData) != 1 {
+		return "", fmt.Errorf("secret has many fields")
+	}
+
+	// Expect htpasswd field is present.
+	htpasswdBytes, ok := secretData["htpasswd"]
+	if !ok {
+		return "", fmt.Errorf("secret has no htpasswd field")
+	}
+	htpasswd := string(htpasswdBytes)
+
+	// Expect only one user-password pair.
+	if strings.Count(htpasswd, "{PLAIN}") != 1 {
+		return "", fmt.Errorf("secret has many users in htpasswd field")
+	}
+
+	userPrefix := defaultUserName + ":{PLAIN}"
+	if strings.Count(htpasswd, userPrefix) != 1 {
+		return "", fmt.Errorf("secret has no password for %s user", defaultUserName)
+	}
+
+	// Extract password.
+	cleaned := strings.TrimSpace(htpasswd)
+	pass := strings.TrimPrefix(cleaned, defaultUserName+":{PLAIN}")
+
+	// Expect password length is sufficient.
+	if len(pass) < generatedPasswdLength {
```

Comment:
Can we also add it to the openapi spec? I didn't get the logic with this length check.

Reply:
Done. `auth.password` is deprecated now, no length check required.